Transfers: Add conveyor-preparer daemon; #4056

bziemons opened the pull request:
- Add preparer daemon
- Searches for minimum distance source and sets QUEUED or WAITING state based on throttler mode and limits.
- Add PREPARING request state.
- Add conveyor.use_preparer config option.
- Add tests for preparer.

cserf commented on lib/rucio/db/sqla/constants.py:
```diff
@@ -1,19 +1,27 @@
-# - Wen Guan, <wen.guan>, 2016
+# - Vincent Garonne <vincent.garonne@cern.ch>, 2015-2017
+# - Wen Guan <wen.guan@cern.ch>, 2015-2016
+# - Cedric Serfon <cedric.serfon@cern.ch>, 2016
```
Why do you change the dates ? Is it based on git blame ?

bziemons replied:
its based on tools/add_header and therefore the git history - is there a problem with these lines? The history isn't always correct, especially when files were moved around

bari12 replied:
The constants file was probably (content) copied once when we move to the sqla folder. Sometimes the dates in these files are also wrongly inserted by people. I think it is fine to just rely on the add_header one.

cserf commented on lib/rucio/core/transfer.py:
```diff
@@ -1260,8 +1258,8 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
-def __list_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
-                                                 limit=None, activity=None, older_than=None, rses=None, session=None):
+def __list_transfer_requests_and_source_replicas(total_workers=0, worker_number=0, limit=None, activity=None,
+                                                 older_than=None, rses=None, request_state=None, session=None):
```
This method is called by `get_transfer_requests_and_source_replicas`, but you changed the signature by introducing a request_state. The call should be changed in `get_transfer_requests_and_source_replicas`

bziemons replied:
will do this anyway, although it works fine with the default

bziemons replied:
just saw this `request_state=RequestState.QUEUED` in the `get_transfer_requests_and_source_replicas` call …? Nothing to do here.

cserf commented on lib/rucio/daemons/conveyor/preparer.py:
```diff
@@ -0,0 +1,224 @@
+    return f'updated {count}/{limit} requests'
+
+
+@transactional_session
```
Shouldn't it be moved to core ?

bziemons replied:
Good question. It is preparer-specific though - I feel like it would be misplaced in core.

bari12 replied:
It's debatable, but in general we tried to put everything with database interactions to the core, even if only used by the one daemon. (It's the same for the judges etc.)
I would move it.

bziemons replied:
will do

cserf commented on lib/rucio/daemons/conveyor/preparer.py:
```diff
@@ -0,0 +1,224 @@
+        return 'had nothing to do'
+
+    # sort by Request.id, should be pretty quick since the database should return it this way (tim sort)
+    req_sources = sorted(req_sources, key=lambda t: t[request_id_col])
```
What type is `req_sources` ?

bziemons replied:
list

bziemons replied:
Replaced `sorted` with `list.sort`.

cserf commented on lib/rucio/core/transfer.py:
```diff
@@ -1273,6 +1271,10 @@ def __list_transfer_requests_and_source_replicas(total_workers=0, worker_number=
+
+    if request_state is None:
+        request_state = RequestState.QUEUED
+
```
Ah, sorry, forgot my comment above :D

cserf reviewed (commented):
A first few comments

bziemons committed:
Transfers: Add preparer daemon; Fix #4056

Searches for minimum distance source and sets QUEUED or WAITING state
based on throttler mode and limits.
Add PREPARING request state.
Add conveyor.use_preparer config option.
Add tests for preparer.
